### PR TITLE
gh-91520: Rewrite imghdr inlining for clarity and completeness

### PR DESCRIPTION
**Pull request by warsaw**
* Move MIMEImage class back closer to the top of the file since it's the
  important thing.
* Use a decorate to mark a given rule function and simplify the rule function
  names for clarity.
* Copy over all the imghdr test data files into the email package's test data
  directory.  This way when imghdr is actually removed, it won't affect the
  MIMEImage guessing tests.
* Rewrite and extend the MIMEImage tests to test for all supported
  auto-detected MIME image subtypes.
* Remove the now redundant PyBanner048.gif data file.

Automerge-Triggered-By: GH:warsaw

#91520

### CODE REVIEW

**Comment by warsaw** [comment]
@ngnpope

**Comment by arhadthedev** [comment]
> ```
> # Open the files in binary mode.  Use imghdr to figure out the
> # MIME subtype for each specific image.
> ```

Probably, this mention of `imghdr` should be removed too.

File: Lib/email/mime/image.py [hunk 1]
```diff
@@ -10,137 +10,143 @@
...
 # Originally from the imghdr module.
-def _what(h):
-    for tf in tests:
-        if res := tf(h):
+def _what(data):
+    for rule in _rules:
+        if res := rule(data):
             return res
     else:
         return None
```
Comment (arhadthedev) [inline Lib/email/mime/image.py:55]:
```suggestion
    verdicts = (rule(data) for rule in _rules)
    recognized_formats = (format for format in verdicts if format is not None)
    return next(recognized_formats, None)
```

We can go functional and describe "what" we want to get, not "how".

File: Lib/email/mime/image.py [hunk 2]
```diff
@@ -10,137 +10,143 @@
...
+        Content-Transfer-Encoding or other headers to the message as
+        necessary.  The default encoding is Base64.
+
+        Any additional keyword arguments are passed to the base class
+        constructor, which turns them into parameters on the Content-Type
+        header.
+        """
+        if _subtype is None:
+            if (_subtype := _what(_imagedata)) is None:
+                raise TypeError('Could not guess image MIME subtype')
```
Comment (arhadthedev) [inline Lib/email/mime/image.py]:
```suggestion
        _subtype = _what(_imagedata) if _subtype is None else _subtype
        if _subtype is None:
            raise TypeError('Could not guess image MIME subtype')
```

File: Lib/email/mime/image.py [hunk 3]
```diff
@@ -10,137 +10,143 @@
...
+    for rule in _rules:
+        if res := rule(data):
             return res
     else:
         return None
 
-tests = []
 
-def _test_jpeg(h):
+def rule(rulefunc):
```
Comment (brettcannon) [inline Lib/email/mime/image.py:58]:
```suggestion
def _rule(rulefunc):
```

Do you want to expose this publicly?

File: Lib/email/mime/image.py [hunk 4]
```diff
@@ -10,137 +10,143 @@
...
+    _rules.append(rulefunc)
+    return rulefunc
+
+
+@rule
+def jpeg(h):
```
Comment (brettcannon) [inline Lib/email/mime/image.py]:
```suggestion
def _jpeg(h):
```

Do you want the name to be public? Same question for the other functions.

File: Lib/email/mime/image.py (hunk 3 above)
Comment (brettcannon) [inline Lib/email/mime/image.py:58]: I _think_ all of these functions consist of checks that could be a single expression. As such, another option is a dict keyed on the format name and a lambda that tests the data as the value.

Comment (warsaw) [reply Lib/email/mime/image.py:58]: I thought about that and originally did go with `_rule()` for the decorator name.  I ended up using `rule()` for a couple of reasons:
* `@rule` is more readable
* This leaves the option to advertise `@email.mime.image.rule` for third party extensions in the future (I mean, third parties could do that now but we don't need to document it).
* `__all__` only names `MIMEImage`

File: Lib/email/mime/image.py (hunk 4 above)
Comment (warsaw) [reply Lib/email/mime/image.py]: For these `__all__` should be enough of an indication these are public, but the other reasons are less relevant for these rule functions.  I went back and forth on it and ultimate just didn't like the underscore names.  That said, I'm on the fence.

Comment (warsaw) [reply Lib/email/mime/image.py]: Okay, I guess `_foo()` isn't too bad.

File: Lib/email/mime/image.py (hunk 3 above)
Comment (warsaw) [reply Lib/email/mime/image.py:58]: They could be, yeah, but it looks less readable to my eyes.

File: Lib/email/mime/image.py (hunk 1 above)
Comment (warsaw) [reply Lib/email/mime/image.py:55]: Yep, thanks for the suggestion, but I think I'll keep the for loop.  Nothing wrong with those and it seems more readable to me.

**Pull request merged**